zebra-chain/src/orchard/keys.rs

```diff
@@ -0,0 +1,1008 @@
+    type Err = SerializationError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match bech32::decode(s) {
+            Ok((hrp, bytes, Variant::Bech32)) => {
+                let decoded = Vec::<u8>::from_base32(&bytes).unwrap();
+
+                let mut scalar_bytes = [0u8; 32];
+                scalar_bytes[..].copy_from_slice(&decoded[0..32]);
+
+                Ok(IncomingViewingKey {
+                    network: match hrp.as_str() {
+                        ivk_hrp::MAINNET => Network::Mainnet,
+                        _ => Network::Testnet,
+                    },
+                    scalar: pallas::Scalar::from_bytes(&scalar_bytes).unwrap(),
+                })
+            }
+            _ => Err(SerializationError::Parse("bech32 decoding error")),
+        }
+    }
+}
+
+impl PartialEq<[u8; 32]> for IncomingViewingKey {
+    fn eq(&self, other: &[u8; 32]) -> bool {
+        self.scalar.to_bytes() == *other
+    }
+}
+
+/// Magic human-readable strings used to identify what networks Orchard full
+/// viewing keys are associated with when encoded/decoded with bech32.
+///
+/// https://zips.z.cash/protocol/nu5.pdf#orchardfullviewingkeyencoding
+mod fvk_hrp {
+    pub const MAINNET: &str = "zviewo";
+    pub const TESTNET: &str = "zviewtestorchard";
+}
+
+/// Full Viewing Keys
+///
+/// Allows recognizing both incoming and outgoing notes without having
+/// spend authority.
+///
+/// For incoming viewing keys on the production network, the
+/// Human-Readable Part is “zviewo”. For incoming viewing keys on the
+/// test network, the Human-Readable Part is “zviewtestorchard”.
+///
+/// https://zips.z.cash/protocol/protocol.pdf#orchardfullviewingkeyencoding
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct FullViewingKey {
+    network: Network,
+    spend_validating_key: SpendValidatingKey,
+    nullifier_deriving_key: NullifierDerivingKey,
+    ivk_commit_randomness: IvkCommitRandomness,
+}
+
+// TODO: impl a From that accepts a Network?
+
+impl fmt::Debug for FullViewingKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("FullViewingKey")
+            .field("network", &self.network)
+            .field("spend_validating_key", &self.spend_validating_key)
+            .field("nullifier_deriving_key", &self.nullifier_deriving_key)
+            .field("ivk_commit_randomness", &self.ivk_commit_randomness)
+            .finish()
+    }
+}
+
+impl fmt::Display for FullViewingKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut bytes = io::Cursor::new(Vec::new());
+
+        let _ = bytes.write_all(&<[u8; 32]>::from(self.spend_validating_key));
+        let _ = bytes.write_all(&<[u8; 32]>::from(self.nullifier_deriving_key));
+        let _ = bytes.write_all(&<[u8; 32]>::from(self.ivk_commit_randomness));
+
+        let hrp = match self.network {
+            Network::Mainnet => fvk_hrp::MAINNET,
+            _ => fvk_hrp::TESTNET,
```

Comment:
Let's avoid wildcard matches, in case we add more networks in future.
(`Transaction` wildcards caused a bunch of trouble with `Transaction::V5`.)

```suggestion
            Network::Mainnet => fvk_hrp::MAINNET,
            Network::Testnet => fvk_hrp::TESTNET,
```